### PR TITLE
feat(mattermost): add support for multiple channel and DMs IDs

### LINKED ISSUE #3100: [Feature]: [Mattermost] oncall mode instead of sigle-channel listening mode
### Summary

Mattermost bot can reply all metioned message instead of listening to single channel

### Problem statement

Bot of Mattermost channel in OpenClaw can reply mentioned message in all channels, instead of listening single channel, but ZeroClaw can't do this

### Proposed solution

like OpenClaw, bot can respond to any mentioned message. 
OpenClaw's implementation:  https://github.com/openclaw/openclaw/blob/8306eabf85ea0c08e02fb0e45c697e22e77dd8c6/extensions/mattermost/src/mattermost

### Non-goals / out of scope

_No response_

### Alternatives considered

_No response_

### Acceptance criteria

_No response_

### Architecture impact

_No response_

### Risk and rollback

_No response_

### Breaking change?

No

### Data hygiene checks

- [x] I removed personal/sensitive data from examples, payloads, and logs.
- [x] I used neutral, project-focused wording and placeholders.

### PR DESCRIPTION
**Pull request by titulus**
## Summary

- Base branch target (master for all contributions): master
- Problem: Mattermost listener was effectively single-channel focused and lacked Slack-like wildcard discovery behavior.
- Why it matters: Bot should respond to mentions across all eligible channels/DMs in on-call mode without restarts and without replaying old history.
- What changed: Added multi-channel selectors for Mattermost, wildcard handling, scoped discovery behavior, cursor bootstrap fixes, schema coverage, and docs updates (EN + zh-CN).
- What did not change (scope boundary): No changes to security model, provider stack, or unrelated channel implementations.

## Label Snapshot (required)

- Risk label (risk: low|medium|high): risk: medium
- Size label (size: XS|S|M|L|XL, auto-managed/read-only): auto
- Scope labels (core|agent|channel|config|cron|daemon|doctor|gateway|health|heartbeat|integration|memory|observability|onboard|provider|runtime|security|service|skillforge|skills|tool|tunnel|docs|dependencies|ci|tests|scripts|dev, comma-separated): channel,config,docs,tests,cron,daemon,onboard
- Module labels (<module>: <component>, for example channel: telegram, provider: kimi, tool: shell): channel: mattermost
- Contributor tier label (trusted contributor|experienced contributor|principal contributor|distinguished contributor, auto-managed/read-only; author merged PRs >=5/10/20/50): auto
- If any auto-label is incorrect, note requested correction: N/A

## Change Metadata

- Change type (bug|feature|refactor|docs|security|chore): feature
- Primary scope (runtime|provider|channel|memory|security|ci|docs|multi): channel

## Linked Issue

- Closes #3100
- Related #N/A
- Depends on #N/A
- Supersedes #5162

## Supersede Attribution (required when `Supersedes #` is used)

- Superseded PRs + authors (`#<pr> by @<author>`, one per line):
  - #5162 by @dongs0104
- Integrated scope by source PR (what was materially carried forward): None — implementations arrived at the same foundation independently. #5162 explicitly adopted multi-channel HTTP logic from this PR.
- `Co-authored-by` trailers added for materially incorporated contributors? No
- If `No`, explain why: No direct code carry-over; #5162 built on this PR's approach, not the other way around.
- Trailer format check: N/A

## Validation Evidence (required)

Commands and result summary:
```
cargo fmt --all -- --check
cargo clippy --all-targets -- -D warnings
CARGO_BUILD_JOBS=1 cargo test
```

- Evidence provided (test/log/trace/screenshot/perf): local command output, all checks passed
- If any command is intentionally skipped, explain why: full cargo test ran with CARGO_BUILD_JOBS=1 due local SIGKILL on parallel build

## Security Impact (required)

- New permissions/capabilities? (Yes/No): No
- New external network calls? (Yes/No): No
- Secrets/tokens handling changed? (Yes/No): No
- File system access scope changed? (Yes/No): No
- If any Yes, describe risk and mitigation: N/A

## Privacy and Data Hygiene (required)

- Data-hygiene status (pass|needs-follow-up): pass
- Redaction/anonymization notes: no personal data introduced
- Neutral wording confirmation (use ZeroClaw/project-native labels if identity-like wording is needed): confirmed

## Compatibility / Migration

- Backward compatible? (Yes/No): Yes
- Config/env changes? (Yes/No): Yes
- Migration needed? (Yes/No): No
- If yes, exact upgrade steps: optional adoption of channel_ids and wildcard selectors

## i18n Follow-Through (required when docs or user-facing wording changes)

- i18n follow-through triggered? (Yes/No): Yes
- If Yes, locale navigation parity updated in README*, docs/README*, and SUMMARY.md for supported locales (en, zh-CN, ja, ru, fr, vi)? (Yes/No): No (not required for this API/setup-guide delta)
- If Yes, localized runtime-contract docs updated where equivalents exist (minimum for fr/vi: commands-reference, config-reference, troubleshooting)? (Yes/No/N.A.): N.A.
- If Yes, Vietnamese canonical docs under docs/i18n/vi/** synced and compatibility shims under docs/*.vi.md validated? (Yes/No/N.A.): N.A.
- If any No/N.A., link follow-up issue/PR and explain scope decision: scope limited to Mattermost reference/setup docs and zh-CN parity for touched docs

## Human Verification (required)

What was personally validated beyond CI:

- Verified scenarios: wildcard selector behavior, trimmed selector behavior, discovery bootstrap cursor behavior for new channels
- Edge cases checked: single channel_id wildcard, channel_ids wildcard, mixed selector precedence, backward compat parsing
- What was not verified: network issues during tests weren't emulated/tested

## Side Effects / Blast Radius (required)

- Affected subsystems/workflows: Mattermost channel listener, config schema parsing, docs, related config constructors/tests
- Potential unintended effects: selector precedence misunderstandings if both fields set
- Guardrails/monitoring for early detection: added regression tests and schema contract tests

## Agent Collaboration Notes (recommended)

- Agent tools used (if any): custom mattermost skill (not included into PR)
- Workflow/plan summary (if any): align behavior with existing slack channel, keep backward compatibility, add tests/docs, run required checks
- Verification focus: formatting, lint, full tests, Mattermost-specific regressions
- Confirmation: naming + architecture boundaries followed (AGENTS.md + CONTRIBUTING.md): Yes

## Rollback Plan (required)

- Fast rollback command/path: revert commit 72112fa8
- Feature flags or config toggles (if any): use explicit channel selectors without wildcard to disable discovery behavior
- Observable failure symptoms: missing or duplicate message handling across discovered channels

## Risks and Mitigations

- Risk: behavior drift in channel discovery/cursor initialization
- Mitigation: targeted regression tests for wildcard and cursor bootstrap logic

### CODE REVIEW

**Comment by JordanTheJet** [comment]
## PR Review: Agent Assessment

### Comprehension Summary

**What:** This PR adds multi-channel and DM ID support to the Mattermost channel implementation. It introduces a new `channel_ids: Vec<String>` config field alongside the existing `channel_id: Option<String>`, implements wildcard-based auto-discovery of all accessible channels (including DMs), and updates the polling listener to iterate over multiple channels with per-channel cursors. Config schema, docs (EN + zh-CN), and tests are updated accordingly.

**Why:** The Mattermost listener was single-channel only, requiring restarts and manual config updates to monitor additional channels. This aligns Mattermost with Slack's wildcard discovery behavior and closes #3100.

**Blast radius:** Changes touch `src/channels/mattermost.rs` (core logic), `src/config/schema.rs` (new field), `src/channels/mod.rs` (factory wiring), `src/cron/scheduler.rs`, `src/daemon/mod.rs`, `src/onboard/wizard.rs`, `src/tui/onboarding.rs` (all mechanical passthrough of the new field), and 4 doc files. No security, runtime, or gateway paths are affected. Risk: **medium**.

---

### Phase 2 — Gate Checks

- **Draft status:** Not draft. OK.
- **CI status:** Passing per context. OK.
- **Template completeness:** Fully completed, all required sections present and substantive. OK.
- **PR size:** 450 additions, 74 deletions across 11 files. Reasonable for a feature PR — scope is coherent and focused on one subsystem.
- **Privacy/data hygiene:** No PII, credentials, or identity-specific test data detected. OK.
- **Malicious content:** None detected. OK.
- **Supersedes attribution:** N/A.
- **Language enforcement:** Non-English content is exclusively within `docs/i18n/zh-CN/` (translation-purpose). OK.

### Duplicate/Overlap Scan — IMPORTANT FINDING

[blocking] **PR #5162** ("feat(channel): add Mattermost WebSocket listener with multi-channel support" by @dongs0104) is currently open and overlaps significantly with this PR:

- Both add `channel_ids: Vec<String>` to `MattermostConfig`
- Both implement `scoped_channel_ids()` with wildcard discovery
- Both refactor the polling listener for multi-channel iteration
- PR #5162 additionally adds WebSocket-based listening and login-based auth

These two PRs will conflict if both are merged. @jordanthejet — a consolidation decision is needed. Key differences: #5162 is larger (adds WebSocket + auth modes) while #5261 is more focused (multi-channel polling only with docs). They share the same core approach to channel scoping.

---

### Phase 3 — Code Review

#### Rust Idiom Compliance

The code is clean and idiomatic. Good use of:
- `filter_map` with `then_some` for the dedup/trim pipeline in `scoped_channel_ids()`
- `HashSet` for deduplication while preserving insertion order
- `HashMap::entry().or_insert()` for cursor bootstrap
- No new `unwrap()` calls in production code; `unwrap_or_default()` used appropriately
- `#[allow(clippy::cast_possible_truncation)]` properly scoped for the `u128 -> i64` ms cast
- `#[serde(default)]` on the new Vec field ensures backward compatibility

#### Architectural Alignment

- No new dependencies introduced
- No trait boundary violations — changes are entirely within the `MattermostChannel` impl and `Channel` trait `listen()` method
- Factory wiring in `build_channel_by_id` and `collect_configured_channels` correctly passes the new field through

#### Correctness Analysis

The `scoped_channel_ids()` logic is well-structured:
1. `channel_ids` non-empty with `"*"` entry → None (discover all)
2. `channel_ids` non-empty without wildcard → explicit list (deduped, trimmed)
3. `channel_ids` empty, `channel_id` is `"*"` or empty → None (discover all)
4. `channel_ids` empty, `channel_id` has value → single-channel vec

[suggestion] **Edge case in `scoped_channel_ids`:** If `channel_ids = ["*", "chan-a"]`, the wildcard wins and all explicit IDs are ignored. This is documented behavior, but consider logging a warning when a wildcard is mixed with explicit IDs, as this likely indicates a misconfiguration.

[suggestion] **`discover_channels` pagination:** The Mattermost API endpoint `/api/v4/users/{user_id}/teams/{team_id}/channels` returns all channels the user belongs to without pagination support. The `page`/`per_page` query parameters are accepted but unnecessary for this endpoint. This is harmless but adds dead code — the loop will always break on the first iteration since `batch_len < per_page` (200) will always be true unless the bot is in 200+ channels per team.

[suggestion] **Sequential polling in the listen loop:** With many channels, the sequential `for channel_id in &target_channels` loop could introduce latency. If a bot monitors 50 channels, each 3-second poll cycle sequentially hits 50 HTTP endpoints before sleeping. For a future iteration, consider batching or concurrent requests. Not blocking for this PR — the Slack channel has a similar pattern.

#### Backward Compatibility

Verified: existing configs with only `channel_id` set will continue to work identically. The `#[serde(default)]` on `channel_ids` means it defaults to an empty Vec, and the `scoped_channel_ids()` fallback logic correctly handles this case. No behavior change for existing single-channel users.

#### Regression Risk

- The old `listen()` method used to require `channel_id` with `ok_or_else(...)`. Now both fields empty triggers auto-discovery. This is an intentional behavior change and is documented, but it means an existing config with `channel_id` accidentally omitted will silently start discovering all channels instead of erroring. This is arguably better UX but is a behavior shift worth noting in release notes.
- `ensure_poll_cursor` correctly uses `or_insert` (not `or_insert_with`), so existing cursors for known channels are never overwritten. New channels discovered mid-session get `now_ms` as their cursor, preventing history replay. This is correct.

#### Security & Performance

- **Security:** No security impact identified. No new permissions, no credential handling changes, no attack surface expansion. The auto-discovery endpoint uses existing bot token auth.
- **Performance:** The auto-discovery adds periodic HTTP calls (every 60 seconds) when in wildcard mode. This is a modest network overhead. The sequential per-channel polling could scale poorly with many channels but matches existing patterns in the codebase.

#### Test Coverage

New tests added (good coverage):
- `mattermost_channel_ids_wildcard_enables_discovery` — wildcard in `channel_ids`
- `mattermost_single_channel_wildcard_enables_discovery` — wildcard in `channel_id` with trimming
- `mattermost_channel_ids_deduplicate_explicit_ids` — dedup behavior
- `mattermost_single_channel_id_is_trimmed` — trim on `channel_id`
- `mattermost_ensure_poll_cursor_bootstraps_only_new_channels` — cursor bootstrap logic
- Schema tests: `mattermost_config_toml_backward_compat`, `mattermost_config_toml_accepts_channel_ids`, `mattermost_config_toml_accepts_both_channel_selectors`

[suggestion] Missing test: `channel_ids` with all-whitespace or empty-string entries should produce `None` (discovery mode). The current code handles this (the `filter_map` strips empty entries, and if all are empty, `ids` is empty, falling through to `channel_id`), but a test would document this edge case.

[suggestion] No integration test for `discover_channels()` itself (it requires a mock Mattermost server). Acceptable given CI constraints, but worth noting.

#### i18n Follow-Through

- EN docs updated: `channels-reference.md`, `mattermost-setup.md`. OK.
- zh-CN docs updated: `channels-reference.zh-CN.md`, `mattermost-setup.zh-CN.md`. OK.
- ja, ru, fr, vi: No Mattermost docs exist in these locales, so no parity issue.

#### Documentation Review

Docs are clear and complete. The listening scope rules are consistently documented across all 4 doc files with matching semantics. The field table is updated with correct precedence descriptions.

---

### Verdict: Needs Maintainer Decision

This PR is technically well-implemented, well-tested, and well-documented. The code quality is high, backward compatibility is preserved, and the feature is genuinely useful. Under normal circumstances, this would be **ready for maintainer merge**.

However, the significant overlap with **PR #5162** requires a maintainer consolidation decision before this can proceed. Both PRs implement the same core feature (multi-channel + wildcard discovery) with compatible approaches, but they will conflict at merge time.

**Recommended next steps for @jordanthejet:**
1. Decide which PR to land first (or whether to consolidate them).
2. If this PR is chosen: the suggestions above are non-blocking quality improvements that can be addressed in follow-up.
3. If #5162 is chosen: this PR's docs and zh-CN translations could still be cherry-picked as a supplement.

| Field | Content |
|---|---|
| PR | #5261 — feat(mattermost): add support for multiple channel and DMs IDs |
| Author | @titulus |
| Summary | Adds `channel_ids` config field for multi-channel Mattermost monitoring with wildcard auto-discovery, per-channel poll cursors, and docs (EN + zh-CN) |
| Action | Needs-maintainer-review |
| Reason | Significant overlap with open PR #5162 requires consolidation decision |
| Security/performance | No security impact. Modest network overhead from periodic auto-discovery (60s interval). Sequential polling may scale poorly with many channels. |
| Changes requested | None blocking. 4 suggestions for quality improvement. |
| Architectural notes | No new dependencies. No trait violations. Adds ~150 lines of channel logic + ~80 lines of tests. |
| Tests | CI passing. New unit tests cover scoping, dedup, trimming, cursor bootstrap, and schema compat. |
| Notes | PR #5162 overlaps significantly — both add `channel_ids`, `scoped_channel_ids()`, and multi-channel polling. Maintainer must decide consolidation path before merge. |

**Comment by dongs0104** [comment]
@titulus when I use websocket it doesn't matter I did not change polling behavior yet.
So we can merge together

**Comment by dongs0104** [comment]
@titulus please check PR#5162

i pull ur commit at that branch

File: docs/i18n/zh-CN/setup-guides/mattermost-setup.zh-CN.md [hunk 1]
```diff
@@ -24,19 +24,28 @@ ZeroClaw 通过 REST API v4 原生支持与 Mattermost 集成。这种集成非
 [channels_config.mattermost]
 url = \"https://mm.your-domain.com\"
 bot_token = \"your-bot-access-token\"
-channel_id = \"your-channel-id\"
+channel_id = \"your-channel-id\"              # 可选：单频道回退配置
```
Comment (singlerider) [inline docs/i18n/zh-CN/setup-guides/mattermost-setup.zh-CN.md:27]: What is `channel_id` when you're supporting multiple channels? Would this basically become deprecated? I ask because I have an incoming PR for schema versioning that would handle such a case.

Comment (titulus) [reply docs/i18n/zh-CN/setup-guides/mattermost-setup.zh-CN.md:27]: @singlerider yep. `channel_id` is for backward-compatibility with existing configs. The implementation mirrors Slack's channel selector pattern (precedence: channel_ids > channel_id > discovery), so configurations are consistent and predictable across channel implementations.

Comment (singlerider) [reply docs/i18n/zh-CN/setup-guides/mattermost-setup.zh-CN.md:27]: Can you add a comment to that effect (that you intend deprecating)? I'll handle removal in my PR to drop deprecated fields safely. The behavior for the migration will be: If channel_ids not contains channel_id, append to channel_ids, then get rid of channel_id. I realize I left this comment in the Markdown file, but I'm referring to leaving a comment in the actual prop definition location.

Comment (titulus) [reply docs/i18n/zh-CN/setup-guides/mattermost-setup.zh-CN.md:27]: like this? 
image

**Comment by theonlyhennygod** [comment]
## Agent Review — Already Reviewed, Pending Maintainer Decision

This PR was previously reviewed in detail by @JordanTheJet with a thorough assessment covering all gate checks, code review, and architectural analysis. The review identified a significant overlap with PR #5162 that requires a maintainer consolidation decision.

**Current status:** The review findings remain valid. CI is fully green (all checks passing as of the latest push). No new commits have been pushed since the review was posted. The four `[suggestion]` items (wildcard+explicit mixed logging, pagination dead code, sequential polling, missing whitespace-only edge case test) are non-blocking quality improvements.

No additional agent action needed — this PR is blocked on the #5162 consolidation decision by @jordanthejet.

File: docs/i18n/zh-CN/setup-guides/mattermost-setup.zh-CN.md (hunk 1 above)
Comment (singlerider) [reply docs/i18n/zh-CN/setup-guides/mattermost-setup.zh-CN.md:27]: @titulus Yes, please forgive my delayed reply.

Comment (singlerider) [reply docs/i18n/zh-CN/setup-guides/mattermost-setup.zh-CN.md:27]: Also, there is a merge conflict migration guide (I was working on a huge refactor for these past four days) in the body of #5559. #5517 will be how these sorts of deprecations of fields are handled in the future.

**Comment by Audacity88** [comment]
Quick queue-status note: I checked the current PR state, linked issue #3100, the old overlap with #5162, current `master`, and the newer #5902 branch.

Closing this in favor of #5902. The Mattermost on-call / multi-channel feature is still valid, and #3100 should stay open, but this branch is now dirty against `master` and still targets the pre-workspace-split paths (`src/channels/mattermost.rs`, `src/config/schema.rs`, etc.). #5902 has already re-applied the Mattermost WebSocket + multi-channel work onto the current `crates/zeroclaw-channels` / `crates/zeroclaw-config` layout, and its body/comment history carries attribution for your #5261 `channel_ids` work.

@titulus if you want to keep pushing the focused polling-only slice instead, the useful path would be a fresh PR against the current crate layout. For now, I am treating #5902 as the live vehicle and keeping #3100 open until that work lands or is replaced.
